rethinkdb_exporter.go

```diff
@@ -351,7 +356,19 @@ func main() {
 		log.Fatal("need parameter addr with len > 0 to connect to RethinkDB cluster")
 	}
 
-	exporter := NewRethinkDBExporter(*addr, *auth, *user, *pass, *clusterName, *namespace)
+	var err error
+	var tlsConfig *tls.Config
+	if *tlsEnable {
+		tlsConfig, err = prepareTLSConfig(*caFile, *certFile, *keyFile)
+		if err != nil {
+			log.Fatalf("failed to prepare tls config: %v", err)
+		}
+	}
+
+	rconn := connectRethinkdb(*addr, *auth, *user, *pass, tlsConfig)
+	defer rconn.Close()
+
+	exporter := NewRethinkDBExporter(*clusterName, *namespace, rconn)
```

Comment:
I'd prefer you pass (and use) the `r.ConnectOpts` struct into the exporter and have the exporter open the connection (and close it when done scraping).

Reply:
But that's very inefficient, it will lead to open N connections to db every second. Assuming that N is connection pool size in the driver and second is Prometheus pulling rate. In some heavy loaded networks (especially docker) DNS lookup can last for seconds. It will lead to Prometheus timeouts.

Reply:
For me this is less about efficiency but about isolation of scrapes and not require the exporter to have a connection open to the DB and make it clear where a connection starts and ends (plus have the error reporting around having a connection attempt fail).

This should open 1 connection per rethink instance scraped.
Connecting to a DB shouldn't take long and scrape timeouts by default are 10s so that gives you a lot to work with before Prometheus times out.

Not 100% sure if DNS lookups would be cached by the exporter but in general, if you have multi-second DNS lookups in your system then you have other issues you should address first.

Reply:
@oliver006 what exactly confuses you in the concept of connection reusage?

Reply:
It doesn't confuse me, i'm just suggesting you don't do re-use connections but establish a new one for each scrape.

See e.g. the mysql exporter or the redis exporter following a similar pattern:
https://github.com/prometheus/mysqld_exporter/blob/72c32a449e963509045fe43e8fb65cbcf8bccc00/collector/exporter.go#L134-L140

https://github.com/oliver006/redis_exporter/blob/f433cfc9a5e5b77ae9fa10da1d48f08502be650f/exporter.go#L1133-L1139